### PR TITLE
SoundPlayer: Keep track of the selected visualization in the config

### PR DESCRIPTION
**Pull request by cbyrneee**
Before, the SoundPlayer would not keep track of what visualization the user had selected. So, if the user selected a certain visualization (e.g. album cover), and restarted SoundPlayer, it would return to the default (which is bars).

Now, we use `LibConfig` to keep track of the selected visualization, which is a nice UX improvement in my opinion :^)

https://user-images.githubusercontent.com/71222289/226192589-9a962957-2108-45fd-ac49-61e666808cfb.mp4

### CODE REVIEW

**Comment by cbyrneee** [comment]
I'm very open to criticism on this code, as I'm not sure what the best pattern was here for an "enum"-like configuration value. I saw that the browser used a string for the ColorScheme configuration value, so I decided to stay in-line with that.

File: Userland/Applications/SoundPlayer/SoundPlayerWidgetAdvancedView.cpp [hunk 1]
```diff
@@ -53,7 +54,18 @@ SoundPlayerWidgetAdvancedView::SoundPlayerWidgetAdvancedView(GUI::Window& window
     m_volume_icon = Gfx::Bitmap::load_from_file("/res/icons/16x16/audio-volume-medium.png"sv).release_value_but_fixme_should_propagate_errors();
     m_muted_icon = Gfx::Bitmap::load_from_file("/res/icons/16x16/audio-volume-muted.png"sv).release_value_but_fixme_should_propagate_errors();
 
-    m_visualization = m_player_view->add<BarsVisualizationWidget>();
+    auto visualization_widget = Config::read_string("SoundPlayer"sv, "Preferences"sv, "VisualizationWidget"sv, "bars"sv);
+    if (visualization_widget == "samples")
+        m_visualization = m_player_view->add<SampleWidget>();
```
Comment (linusg) [inline Userland/Applications/SoundPlayer/SoundPlayerWidgetAdvancedView.cpp:59]: The naming is the wrong way around IMO, the string is "visualization", and the widget is, well, "visualization widget" :)
I think for now `visualization_widget` -> `visualization` is fine.

Comment (linusg) [reply Userland/Applications/SoundPlayer/SoundPlayerWidgetAdvancedView.cpp:59]: (and `VisualizationWidget` -> `Visualization` for the config name)

File: Userland/Applications/SoundPlayer/SoundPlayerWidgetAdvancedView.cpp [hunk 2]
```diff
@@ -53,7 +54,18 @@ SoundPlayerWidgetAdvancedView::SoundPlayerWidgetAdvancedView(GUI::Window& window
...
+    if (visualization_widget == "album_cover")
+        m_visualization = m_player_view->add<AlbumCoverVisualizationWidget>([this]() {
+            return this->get_image_from_music_file();
+        });
```
Comment (linusg) [inline Userland/Applications/SoundPlayer/SoundPlayerWidgetAdvancedView.cpp:63]: https://github.com/SerenityOS/serenity/blob/master/Documentation/CodingStyle.md#omission-of-curly-braces-from-statement-blocks

File: Userland/Applications/SoundPlayer/SoundPlayerWidgetAdvancedView.cpp [hunk 3]
```diff
@@ -53,7 +54,18 @@ SoundPlayerWidgetAdvancedView::SoundPlayerWidgetAdvancedView(GUI::Window& window
     m_volume_icon = Gfx::Bitmap::load_from_file("/res/icons/16x16/audio-volume-medium.png"sv).release_value_but_fixme_should_propagate_errors();
     m_muted_icon = Gfx::Bitmap::load_from_file("/res/icons/16x16/audio-volume-muted.png"sv).release_value_but_fixme_should_propagate_errors();
 
-    m_visualization = m_player_view->add<BarsVisualizationWidget>();
+    auto visualization_widget = Config::read_string("SoundPlayer"sv, "Preferences"sv, "VisualizationWidget"sv, "bars"sv);
+    if (visualization_widget == "samples")
+        m_visualization = m_player_view->add<SampleWidget>();
+    if (visualization_widget == "album_cover")
+        m_visualization = m_player_view->add<AlbumCoverVisualizationWidget>([this]() {
+            return this->get_image_from_music_file();
```
Comment (linusg) [inline Userland/Applications/SoundPlayer/SoundPlayerWidgetAdvancedView.cpp:62]: omit `this->`

File: Userland/Applications/SoundPlayer/SoundPlayerWidgetAdvancedView.cpp [hunk 4]
```diff
@@ -53,7 +54,18 @@ SoundPlayerWidgetAdvancedView::SoundPlayerWidgetAdvancedView(GUI::Window& window
...
+        });
+
+    // Either the "VisualizationWidget" configuration value is invalid, or it was set to the default value ("bars").
+    if (m_visualization == nullptr) {
```
Comment (linusg) [inline Userland/Applications/SoundPlayer/SoundPlayerWidgetAdvancedView.cpp:66]: That should be an `else`, after previous `if`/`else if`s, IMO.

**Review by linusg — CHANGES_REQUESTED** [review]

File: Userland/Applications/SoundPlayer/SoundPlayerWidgetAdvancedView.cpp [hunk 5]
```diff
@@ -50,7 +54,16 @@ SoundPlayerWidgetAdvancedView::SoundPlayerWidgetAdvancedView(GUI::Window& window
     m_volume_icon = Gfx::Bitmap::load_from_file("/res/icons/16x16/audio-volume-medium.png"sv).release_value_but_fixme_should_propagate_errors();
     m_muted_icon = Gfx::Bitmap::load_from_file("/res/icons/16x16/audio-volume-muted.png"sv).release_value_but_fixme_should_propagate_errors();
 
-    m_visualization = m_player_view->add<BarsVisualizationWidget>();
+    auto visualization = Config::read_string("SoundPlayer"sv, "Preferences"sv, "VisualizationWidget"sv, "bars"sv);
```
Comment (linusg) [inline Userland/Applications/SoundPlayer/SoundPlayerWidgetAdvancedView.cpp:57]: > (and `VisualizationWidget` -> `Visualization` for the config name)

Comment (cbyrneee) [reply Userland/Applications/SoundPlayer/SoundPlayerWidgetAdvancedView.cpp:57]: Oopsies, brushed over that, thank you :)

**Review by linusg — APPROVED** [review]

**Comment by cbyrneee** [comment]
Seems like the on-target tests are taking a suspiciously long time with no log output on my side...

**Pull request merged**